### PR TITLE
Surge-XT: unstable-2021-11-07 -> unstable-2021-12-11

### PR DESCRIPTION
**Pull request by magnetophon**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by symphorien** [comment]
```
[nix-shell:~/.cache/nixpkgs-review/pr-149487]$ $buildInputs/bin/Surge\ XT
Segmentation fault (core dumped)
```

**Comment by baconpaul** [comment]
Looks like you are building the VST3, Standalone and LV2 all from the community LV2 branch
We have applied patches to surge in our repo version which fix core dumps on the standalone which are in juce 6.1.3 (which we haven't certified against), in our surge fork (which we have) but not in the LV2 community.
FYI

**Comment by baconpaul** [comment]
The juce patch which fixes this is the juce repo commit 478f074186f6d0488b9dfc8b8e90a316682ad599

**Comment by baconpaul** [comment]
(or, just build the LV2 against community and build the VST3 and Standalone against our branch, which is where we tested)

**Comment by magnetophon** [comment]
@baconpaul I thought the crash you patched happened at a certain user interaction, not at startup.  In any case, the standalone built with your juce also crashes at startup for me.

**Comment by baconpaul** [comment]
Oh wow well that’s bad! I asked kott to patch lv2 branch anyway
So we’ve been starting the standalone a bunch with no crash. What’s the environment you are using? Maybe we move this to surge repo so as to not clutter nixos tho

**Comment by baconpaul** [comment]
And the environment where you run it has an x display and stuff right?

**Comment by symphorien** [comment]
it's i3+xfce on good old x.

**Comment by baconpaul** [comment]
Does the built asset work if you run it on a desktop proper?
My guess is we are looking for desktop size and getting a crash

**Comment by symphorien** [comment]
what do you mean by desktop proper ?

**Comment by baconpaul** [comment]
oh sorry I thought you said xfvb. xfce is a window manager?
hmm well i wonder why it crashes on your environment and not in any of ours.
Short of installing nixos, any ideas on how we would reproduce this?

**Comment by magnetophon** [comment]
@baconpaul Thanks. 
Unfortunately, I don't know how to reproduce this.
Can you let me know when the juce patch has landed, so I can update this PR?

@symphorien Shall I remove the standalone from the pkg?

**Comment by baconpaul** [comment]
Kott updated the lv2 branch to 6.1.3 last night. We haven't tested against 6.1.3 but it does include the standalone fix patch.

The standalone is super useful for almost everyone except your use case. Are we sure it crashes on all nixos installs vs just the one test case here?

**Comment by symphorien** [comment]
> Shall I remove the standalone from the pkg?

This does not seem to be a regression, so I'll let you decide, as a maintainer. Actually I don't know what surge is, so I can't tell what is best :)

**Comment by magnetophon** [comment]
> Are we sure it crashes on all nixos installs vs just the one test case here?

Forwarding the question to other nixos musicians:
@astro @orivej @OPNA2608 @nico202 @minijackson @puffnfresh @MarcWeber @hirenashah @lovek323 @kira-bruneau @mausch @Moredread @nathyong

File: pkgs/applications/audio/surge-XT/default.nix
Comment (OPNA2608) [inline pkgs/applications/audio/surge-XT/default.nix:53]: This is enabled by default with CMake.
```suggestion
```

File: pkgs/applications/audio/surge-XT/default.nix
Comment (OPNA2608) [inline pkgs/applications/audio/surge-XT/default.nix:57]: The way this is written does not value the user's configured `nix.buildCores`, or any of your `cmakeFlags`. I was abit started when my system suddenly slowed down, because it consumed all 48 cores instead of just the 6 I configured Nix to use by default. Please try moving the targets into `cmakeFlags`.
```suggestion
```

File: pkgs/applications/audio/surge-XT/default.nix
Comment (OPNA2608) [inline pkgs/applications/audio/surge-XT/default.nix:61]: Should be run by default.
```suggestion
```

File: pkgs/applications/audio/surge-XT/default.nix
Comment (OPNA2608) [inline pkgs/applications/audio/surge-XT/default.nix:48]: See the comment on `buildPhase`.
```suggestion
    "surge-xt_LV2" "surge-fx_LV2" "surge-xt_Packaged" "surge-fx_Packaged"
```

**Review by OPNA2608 — CHANGES_REQUESTED** [review]
Result of `nixpkgs-review pr 149487` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>surge-XT</li>
  </ul>
</details>

Binaries segfault immediately, inside & outside the review shell. I'm using the Pantheon DE.

File: pkgs/applications/audio/surge-XT/default.nix
Comment (OPNA2608) [reply pkgs/applications/audio/surge-XT/default.nix:48]: These prolly need to go to `makeFlags` instead, sorry.

**Comment by OPNA2608** [comment]
Build emits this, dunno if that's relevant.
```
Generate ttl data for './SurgeXT.so', basename: 'SurgeXT'
SurgeXT LV2
  - Version      : 0.99.git-no-branch.git-no-commit with JUCE 60102
  - Build Info   : 1980-01-01 00:00:00 using GNU-10.3.0
Surge Error [Unable to set up User Directory]
Permission denied: '/homeless-shelter/.Surge XT'
  - Data         : "/usr/share/Surge XT/"
  - User Data    : "/homeless-shelter/.Surge XT"
Writing manifest.ttl... done!
Writing SurgeXT.ttl... done!
Writing presets.ttl...
Saving preset 1/2... done!
surge-xt: Relocating LV2 component
[100%] Built target surge-xt_LV2
```

**Comment by OPNA2608** [comment]
It needs some X-libs added, either via `makeWrapper` or `patchelf`. I checked with strace & saw that it was trying to open afew of them right before crashing. I'll fiddle abit more with this package & make a PR against your branch with the fixes when I'm done.

Bildschirmfoto von 2021-12-11 17-34-07

**Comment by symphorien** [comment]
> Surge Error [Unable to set up User Directory]
> Permission denied: '/homeless-shelter/.Surge XT

`export HOME=$(mktemp -d)`

surge should probably not touch $HOME during make install...

**Comment by symphorien** [comment]
> It needs some X-libs added, either via makeWrapper or patchelf

Sometime explicitly linking them with `NIX_CFLAGS_COMPILE="-lthelib"` also works

**Comment by OPNA2608** [comment]
> > Surge Error [Unable to set up User Directory]
> > Permission denied: '/homeless-shelter/.Surge XT
> 
> `export HOME=$(mktemp -d)`
> 
> surge should probably not touch $HOME during make install...

I think this is just the LV2 plugin being started for something, because the binaries print roughly the same now that they're working:

```
SurgeXT Standalone
  - Version      : 0.99.git-no-branch.git-no-commit with JUCE 60102
  - Build Info   : 1980-01-01 00:00:00 using GNU-10.3.0
  - Data         : "/nix/store/y6y0lwmv8wwx4dv4b59c7srrc11374xy-surge-XT-unstable-2021-12-07/share/surge-xt"
  - User Data    : "/home/bt1cn/Documents/Surge XT"
```

Maybe someone who knows how to use those kinds of audio plugins could check if the plugin works correctly though.

---

> > It needs some X-libs added, either via makeWrapper or patchelf
> 
> Sometime explicitly linking them with `NIX_CFLAGS_COMPILE="-lthelib"` also works

JUCE lists some xlibs [in their code](https://github.com/juce-framework/JUCE/blob/0bac0e78c815d9b6328e238fefc2bb151a5163ec/modules/juce_gui_basics/native/x11/juce_linux_X11_Symbols.h#L601-L614) as `DynamicLibrary`s, I assume this means that it wants to `dlopen` them at runtime. I've fixed the crashing by forcing everything to be dynamically linked against them, I'm not well versed in how JUCE & VST3/LV2 works though.

The standalone binaries should certainly rely on regular linking for them, the plugins/libraries I'm not so sure about.

**Comment by magnetophon** [comment]
> Maybe someone who knows how to use those kinds of audio plugins could check if the plugin works correctly though.

Yes, everything works as expected.
Thanks again!

**Comment by baconpaul** [comment]
> > > Surge Error [Unable to set up User Directory]
> > > Permission denied: '/homeless-shelter/.Surge XT
> > 
> > 
> > `export HOME=$(mktemp -d)`
> > surge should probably not touch $HOME during make install...
> 
> I think this is just the LV2 plugin being started for something, because the binaries print roughly the same now that they're working:
> 
> ```
> SurgeXT Standalone
>   - Version      : 0.99.git-no-branch.git-no-commit with JUCE 60102
>   - Build Info   : 1980-01-01 00:00:00 using GNU-10.3.0
>   - Data         : "/nix/store/y6y0lwmv8wwx4dv4b59c7srrc11374xy-surge-XT-unstable-2021-12-07/share/surge-xt"
>   - User Data    : "/home/bt1cn/Documents/Surge XT"
> ```
> 
> Maybe someone who knows how to use those kinds of audio plugins could check if the plugin works correctly though.
> 
> > > It needs some X-libs added, either via makeWrapper or patchelf
> > 
> > 
> > Sometime explicitly linking them with `NIX_CFLAGS_COMPILE="-lthelib"` also works
> 
> JUCE lists some xlibs [in their code](https://github.com/juce-framework/JUCE/blob/0bac0e78c815d9b6328e238fefc2bb151a5163ec/modules/juce_gui_basics/native/x11/juce_linux_X11_Symbols.h#L601-L614) as `DynamicLibrary`s, I assume this means that it wants to `dlopen` them at runtime. I've fixed the crashing by forcing everything to be dynamically linked against them, I'm not well versed in how JUCE & VST3/LV2 works though.
> 
> The standalone binaries should certainly rely on regular linking for them, the plugins/libraries I'm not so sure about.

One of the problems with LV2 is it needs to run the plugin to build it yeah. And we don't know that we are in the context that we are being probed by the LV2 so we do the normal startup path.

Options are

1. Don't build the LV2, just build the VST3 and Standalone
2. Set the environment variable `XDG_DOCUMENTS_DIR` to a writable directory (any writable directory will work) and surge will use that or
3. Modify the surge code and the LV2 wrapper code to indicate we don't have a home dir some other way

I think 2 is the best choice. Depending on how you do it, could be open to a PR for 3 in the next couple of days as we wrap up the beta. 1 seems kinda not right.

**Comment by magnetophon** [comment]
@baconpaul Thanks for chiming in.
I implemented suggestion nr. 2.

I also updated to todays version.
If everyone involved is happy with the current state, I will squash the commits.

File: pkgs/applications/audio/surge-XT/default.nix
Comment (OPNA2608) [inline pkgs/applications/audio/surge-XT/default.nix:26]: Since this is technically a derivation as well, please update the date in the version above this as well.

Comment (magnetophon) [reply pkgs/applications/audio/surge-XT/default.nix:26]: Sorry, I don't understand, please explain.

Comment (OPNA2608) [reply pkgs/applications/audio/surge-XT/default.nix:26]: https://github.com/NixOS/nixpkgs/blob/089ccb451a3cd8b141999c56edf8807095d57f7f/pkgs/applications/audio/surge-XT/default.nix#L20

needs to be bumped to `2021-12-11`

Comment (magnetophon) [reply pkgs/applications/audio/surge-XT/default.nix:26]: Oh, of course!

**Comment by magnetophon** [comment]
I squashed the commits.

**Comment by magnetophon** [comment]
@OPNA2608 Since you managed to fix the ``BadWindow`` crash in surge, maybe you want to take alook at my WIP for Fire.
Same error, but your fix doesn't help:

```nix
{ lib, stdenv, fetchFromGitHub, libGL, libX11, libXext, libXrender, libXrandr, libXinerama, libXcursor, freetype, alsa-lib, libjack2, curl, lv2, python, cmake, pkg-config, gcc-unwrapped }:
let
  readerwriterqueue = rec {
    version = "1.0.6";
    src = fetchFromGitHub {
      owner = "cameron314";
      repo = "readerwriterqueue";
      rev = "v${version}";
      sha256 = "sha256-g7NX7Ucl5GWw3u6TiUOITjhv7492ByTzACtWR0Ph2Jc=";
    };
  };
  Catch2 = rec {
    version = "2.13.7";
    src = fetchFromGitHub {
      owner = "catchorg";
      repo = "Catch2";
      rev = "v${version}";
      sha256 = "sha256-NhZ8Hh7dka7KggEKKZyEbIZahuuTYeCT7cYYSUvkPzI=";
    };
  };

in

stdenv.mkDerivation rec {
  pname = "Fire";
  version = "0.9.5";

  src = fetchFromGitHub {
    owner = "jerryuhoo";
    repo = pname;
    rev = "v${version}";
    sha256 = "sha256-7I7dezN2gR4rF+Qr6ULHsv2wRlJ/tMrzz0Z/jIm8bTg=";
    fetchSubmodules = true;
  };

  enableParallelBuilding = true;

  nativeBuildInputs = [ cmake pkg-config ];

  buildInputs = [
    libGL libX11 libXext libXrender libXrandr libXinerama libXcursor freetype alsa-lib libjack2 curl lv2 python
  ];

  cmakeFlags = [
    "-DCMAKE_AR=${gcc-unwrapped}/bin/gcc-ar"
    "-DCMAKE_RANLIB=${gcc-unwrapped}/bin/gcc-ranlib"
    "-DCMAKE_NM=${gcc-unwrapped}/bin/gcc-nm"
  ];

  # JUCE dlopen's these at runtime, crashes without them
  NIX_LDFLAGS = (toString [
    "-lXext"
    "-lX11"
    "-lXcursor"
    "-lXinerama"
    "-lXrender"
    "-lXrandr"
  ]);

  postPatch = ''
      cp -R --no-preserve=mode,ownership ${readerwriterqueue.src} readerwriterqueue
      cp -R --no-preserve=mode,ownership ${Catch2.src} Catch2
      sed -i '176, 181d' CMakeLists.txt
      sed -i '33, 46d' CMakeLists.txt
      sed -i '32 i add_subdirectory(JUCE)' CMakeLists.txt
      sed -i '32 i add_subdirectory(readerwriterqueue)' CMakeLists.txt
      sed -i '32 i add_subdirectory(Catch2)' CMakeLists.txt
      sed -i 's/COPY_PLUGIN_AFTER_BUILD TRUE/COPY_PLUGIN_AFTER_BUILD FALSE/' CMakeLists.txt # attempts to write to /homeless-shelter
      patchShebangs .
  '';

  installPhase = ''
    mkdir -p $out/lib/vst3
    cd Fire_artefacts/Release
    cp -r VST3/Fire.vst3 $out/lib/vst3
  '';

  meta = with lib; {
    homepage = "https://github.com/jerryuhoo/Fire";
    description = "This is a distortion plugin developed by Wings!";
    maintainers = [ maintainers.magnetophon ];
    platforms = platforms.linux;
    license = licenses.agpl3;
  };
}

```

**Pull request merged**
